cpp/src/io/orc/dict_enc.cu

```diff
@@ -16,471 +16,265 @@
+                                                        cuco::empty_key{KEY_SENTINEL},
+                                                        cuco::empty_value{VALUE_SENTINEL});
+  auto cur_row          = start_row + t;
+  // all threads should loop over the same number of rows
+  while (cur_row - t < end_row) {
+    auto const is_valid = cur_row < end_row and cur_row < col.size() and col.is_valid(cur_row);
+
+    // insert element at cur_row to hash map and count successful insertions
+    size_type is_unique      = 0;
+    size_type uniq_elem_size = 0;
+
+    if (is_valid) {
+      auto const hash_fn     = hash_functor{col};
+      auto const equality_fn = equality_functor{col};
+      is_unique = hash_map_mutable.insert(std::pair(cur_row, cur_row), hash_fn, equality_fn);
+      if (is_unique) { uniq_elem_size = col.element<string_view>(cur_row).size_bytes(); }
     }
+
+    using block_reduce = cub::BlockReduce<size_type, block_size>;
+    __shared__ typename block_reduce::TempStorage reduce_storage;
+    __shared__ size_type total_num_dict_entries;
+
+    auto const num_unique = block_reduce(reduce_storage).Sum(is_unique);
     __syncthreads();
-    bool collision         = false;
-    uint32_t colliding_row = 0;
-    uint32_t pos_new       = 0;
-    if (i + t < nnz) {
-      pos_new   = s->map.u16[hash];
-      collision = (pos != pos_old && pos_new > pos_old + 1);
-      if (collision) { colliding_row = s->dict[pos_old]; }
-    }
+    auto const char_count = block_reduce(reduce_storage).Sum(uniq_elem_size);
     __syncthreads();
-    if (collision) { atomicMin(s->dict + pos_old, ck_row); }
 
-    __syncthreads();
-    // Resolve collision
-    if (collision && ck_row == s->dict[pos_old]) { s->dict[pos] = colliding_row; }
-  }
-  __syncthreads();
-  // Now that the strings are ordered by hash, compare every string with the first entry in the hash
-  // map, the position of the first string can be inferred from the hash map counts
-  dict_char_count = 0;
-  for (uint32_t i = 0; i < nnz; i += block_size) {
-    uint32_t ck_row = 0, ck_row_ref = 0, is_dupe = 0;
-    if (i + t < nnz) {
-      ck_row                   = s->dict[i + t];
-      string_view string_value = s->chunk.leaf_column->element<string_view>(ck_row + start_row);
-      auto const string_length = static_cast<uint32_t>(string_value.size_bytes());
-      auto const hash          = hash_string(string_value);
-      ck_row_ref               = s->dict[(hash > 0) ? s->map.u16[hash - 1] : 0];
-      if (ck_row_ref != ck_row) {
-        string_view reference_string =
-          s->chunk.leaf_column->element<string_view>(ck_row_ref + start_row);
-        is_dupe = (string_value == reference_string);
-        dict_char_count += (is_dupe) ? 0 : string_length;
-      }
+    if (t == 0) {
+      auto dict_g            = &dictionaries[col_idx][stripe_idx];
+      total_num_dict_entries = atomicAdd(&dict_g->entry_count, num_unique);
+      total_num_dict_entries += num_unique;
```

Comment:
Is this `total_num_dict_entries` used anywhere?

Reply:
good catch! Simplified the kernel logic around this.